### PR TITLE
DAOS-14214 control: Fix potential missed call to drpc failure handlers

### PR DESCRIPTION
**Pull request by tanabarr**
Refactor harness CallDrpc method to simplify and try to improve clarity
of intent. Remove potential for failure handlers not to get called on
a dRPC comms related error. Also provide minor fixes related to review
comments on previously landed PRs 12858 and 12871. 

Features: control
Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/server/harness_test.go [hunk 1]
```diff
@@ -578,6 +563,79 @@ func TestServer_Harness_CallDrpc(t *testing.T) {
...
+					Started:     atm.NewBool(true),
+					CallDrpcErr: errEngineNotReady,
+				},
+				{
+					Ready:       atm.NewBool(true),
+					CallDrpcErr: errors.New("whoops"),
+				},
+			},
+			expErr: errors.New("whoops"),
+		},
```
Comment (kjacque) [inline src/control/server/harness_test.go:594]: Is this equivalent to the case removed above?

**Review by kjacque — APPROVED** [review]

File: src/control/server/harness_test.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/harness_test.go:594]: yes, covers the same case

File: src/control/server/harness.go [hunk 2]
```diff
@@ -156,44 +158,61 @@ func (h *EngineHarness) OnDrpcFailure(fns ...func(ctx context.Context, err error
...
 		}
-		// If the context was canceled, don't trigger callbacks.
-		if errors.Cause(err) == context.Canceled {
-			return
+
+		var e error
+		hasDRPCErr := false
+		for _, e = range drpcErrs {
+			switch e {
+			case errDRPCNotReady, FaultDataPlaneNotStarted:
```
Comment (knard-intel) [inline src/control/server/harness.go:172]: Please could you tell me why `errEngineNotReady` is not included as it is not a DPC a error.
From what I understand, it should be more  `errDRPCNotReady` which should not be included (as it was before).
However, I am not sure if such error should be managed by `onDrpcFailure` as it seems to not really be a DRPC error.

**Review by knard-intel — COMMENTED** [review]
One minor point not clear to me.

**Review by kjacque — APPROVED** [review]

File: src/control/server/harness.go (hunk 2 above)
Comment (tanabarr) [reply src/control/server/harness.go]: This line describes errors that should trigger a leadership election which is needed when engines can no longer can be communicated with. `errEngineNotReady` is not in this list as it's typically associated with start-up related temporal un-availability. `errEngineNotReady` is in the retry list however.

Comment (knard-intel) [reply src/control/server/harness.go]: OK, thanks for the explanation.

**Review by knard-intel — APPROVED** [review]

**Review by mjmac — APPROVED** [review]

**Pull request merged**